File: codeblock/dict.py
# Dict 字面量

# 创建
# python 3.4 不支持 e = { **d } 的方法来创建
# with None as key and as value

d = { 'name' : 'john', 1 : 26, 'age' : 20, None: None }

assert d['name'] == 'john'
assert d[1] == 26
assert d['age'] == 20
assert d[None] == None

assert len(d) == 4


## nested dict

d = {
     0: {
         'a': 1,
         'b': 2,
        },
     1: {
         'c': 3,
         'd': 4,
        },
     None: {
         None: None,
        }
    }

assert d[0]['a'] == 1
assert d[0]['b'] == 2
assert d[1]['c'] == 3
assert d[1]['d'] == 4
assert d[None][None] == None


# 操作符

## ==
# not support
#assert d == {}


# 方法

## clear

d.clear()

assert len(d) == 0


## copy

d = { 'name' : 'john', 1 : 26, 'age' : 20, None: None }
m = d.copy()

d['name'] = 'jane'
d[None] = 'nil'
m['age'] = 56

assert d['age'] == 20
assert d[None] == 'nil'
assert m['name'] == 'john'
assert m[None] == None


## fromkeys

m = d.fromkeys([1, 2, 'height', None])

assert m[1] == None
assert m[2] == None
assert m['height'] == None
assert m[None] == None

m = d.fromkeys([1, 2, 'height', None], 40)

assert m[1] == 40
assert m[2] == 40
assert m['height'] == 40
assert m[None] == 40


## get

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

assert d.get('name') == 'john'
assert d.get('home') == None
assert d.get('home', 'cn') == 'cn'


## items

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

for k, v in d.items():
    if k == 'name':
        assert v == 'john'
    elif k == 1:
        assert v == 26
    elif k == 'age':
        assert v == 20


## keys

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

for k in d.keys():
    assert k == 'name' or k == 1 or k == 'age'


## pop

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

assert d.pop('name') == 'john'
# raise KeyError in python
#assert d.pop('name') == None
assert d.pop('name', 'jane') == 'jane'


## popitem

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

k, v = d.popitem()
if k == 'name':
    assert v == 'john'
elif k == 1:
    assert v == 26
elif k == 'age':
    assert v == 20


## setdefault

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

v = d.setdefault('name', 'bing')
assert v == 'john'
assert d['name'] == 'john'

v = d.setdefault('hair', 'black')
assert v == 'black'
assert d['hair'] == 'black'


## update

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

d.update({ 'age' : 22, 'hair' : 'blond' })
assert d['age'] == 22
assert d['hair'] == 'blond'


## values

d = { 'name' : 'john', 1 : 26, 'age' : 20 }

for v in d.values():
    assert v == 'john' or v == 26 or v == 20

File: tests/dict.lua
local d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20), [_to_null(nil)] = _to_null(nil)})
assert(bool({}, (d[_to_null("name")] == "john")))
assert(bool({}, (d[_to_null(1)] == 26)))
assert(bool({}, (d[_to_null("age")] == 20)))
assert(bool({}, (d[_to_null(nil)] == nil)))
assert(bool({}, (len(merge_kwargs({}, {}), d) == 4)))
d = dict({}, {[_to_null(0)] = _to_null(dict({}, {[_to_null("a")] = _to_null(1), [_to_null("b")] = _to_null(2)})), [_to_null(1)] = _to_null(dict({}, {[_to_null("c")] = _to_null(3), [_to_null("d")] = _to_null(4)})), [_to_null(nil)] = _to_null(dict({}, {[_to_null(nil)] = _to_null(nil)}))})
assert(bool({}, (d[_to_null(0)][_to_null("a")] == 1)))
assert(bool({}, (d[_to_null(0)][_to_null("b")] == 2)))
assert(bool({}, (d[_to_null(1)][_to_null("c")] == 3)))
assert(bool({}, (d[_to_null(1)][_to_null("d")] == 4)))
assert(bool({}, (d[_to_null(nil)][_to_null(nil)] == nil)))
d.clear(merge_kwargs({}, {}))
assert(bool({}, (len(merge_kwargs({}, {}), d) == 0)))
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20), [_to_null(nil)] = _to_null(nil)})
local m = d.copy(merge_kwargs({}, {}))
d[_to_null("name")] = "jane"
d[_to_null(nil)] = "nil"
m[_to_null("age")] = 56
assert(bool({}, (d[_to_null("age")] == 20)))
assert(bool({}, (d[_to_null(nil)] == "nil")))
assert(bool({}, (m[_to_null("name")] == "john")))
assert(bool({}, (m[_to_null(nil)] == nil)))
m = d.fromkeys(merge_kwargs({}, {}), list({}, {_to_null(1, 2, "height", nil)}))
assert(bool({}, (m[_to_null(1)] == nil)))
assert(bool({}, (m[_to_null(2)] == nil)))
assert(bool({}, (m[_to_null("height")] == nil)))
assert(bool({}, (m[_to_null(nil)] == nil)))
m = d.fromkeys(merge_kwargs({}, {}), list({}, {_to_null(1, 2, "height", nil)}), 40)
assert(bool({}, (m[_to_null(1)] == 40)))
assert(bool({}, (m[_to_null(2)] == 40)))
assert(bool({}, (m[_to_null("height")] == 40)))
assert(bool({}, (m[_to_null(nil)] == 40)))
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
assert(bool({}, (d.get(merge_kwargs({}, {}), "name") == "john")))
assert(bool({}, (d.get(merge_kwargs({}, {}), "home") == nil)))
assert(bool({}, (d.get(merge_kwargs({}, {}), "home", "cn") == "cn")))
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
for _, k, v in d.items(merge_kwargs({}, {})) do
    if bool({}, (k == "name")) then
        assert(bool({}, (v == "john")))
    elseif bool({}, (k == 1)) then
        assert(bool({}, (v == 26)))
    elseif bool({}, (k == "age")) then
        assert(bool({}, (v == 20)))
    end
    ::loop_label_1::
end
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
for _, k in d.keys(merge_kwargs({}, {})) do
    assert(bool({}, ((k == "name") or (k == 1) or (k == "age"))))
    ::loop_label_2::
end
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
assert(bool({}, (d.pop(merge_kwargs({}, {}), "name") == "john")))
assert(bool({}, (d.pop(merge_kwargs({}, {}), "name", "jane") == "jane")))
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
local k, v = d.popitem(merge_kwargs({}, {}))
if bool({}, (k == "name")) then
    assert(bool({}, (v == "john")))
elseif bool({}, (k == 1)) then
    assert(bool({}, (v == 26)))
elseif bool({}, (k == "age")) then
    assert(bool({}, (v == 20)))
end
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
local v = d.setdefault(merge_kwargs({}, {}), "name", "bing")
assert(bool({}, (v == "john")))
assert(bool({}, (d[_to_null("name")] == "john")))
v = d.setdefault(merge_kwargs({}, {}), "hair", "black")
assert(bool({}, (v == "black")))
assert(bool({}, (d[_to_null("hair")] == "black")))
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
d.update(merge_kwargs({}, {}), dict({}, {[_to_null("age")] = _to_null(22), [_to_null("hair")] = _to_null("blond")}))
assert(bool({}, (d[_to_null("age")] == 22)))
assert(bool({}, (d[_to_null("hair")] == "blond")))
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
for _, v in d.values(merge_kwargs({}, {})) do
    assert(bool({}, ((v == "john") or (v == 26) or (v == 20))))
    ::loop_label_3::
end
d = dict({}, {[_to_null("name")] = _to_null("john"), [_to_null(1)] = _to_null(26), [_to_null("age")] = _to_null(20)})
assert(bool({}, (operator_in("name", d))))
assert(bool({}, (not operator_in("height", d))))
return {
    d = d,
    m = m,
    k, v = k, v,
    v = v,
}
